### PR TITLE
[doc] Clarify our system and software requirements

### PR DESCRIPTION
**Pull request by imphil**
Before, we didn't explicitly list our system requirements in a 
distribution-agnostic way. Instead, we provided a list of packages and a 
reference distribution (Ubuntu 16.04). Users of other distributions then 
had to either try, or manully cross-check if their setup works as expected.

With this change the documentation describes better what packages are 
required, and which version is required. This commit is meant to be a 
clarification without changing the status quo.

Notes:
* The list of package dependencies and their versions isn't bullet-proof,
  as packages can change behavior even in-between minor
  versions (if they use semantic versioning at all). The only way to be
  sure that a distribution works is testing it. We currently do all our
  tests on Ubuntu 16.04, hence we call this the reference platform.
* OpenTitan has generally been doing a very good job at keeping
  compatibility with as many distributions and setups as
  possible, and we have been very open to feedback to improve the
  development experience for our users. We will continue to do so, but
  need user input for this: we cannot provide automated testing for all
  possible setups.

### CODE REVIEW

**Comment by imphil** [comment]
@martin-lueker @rasmus-madsen This PR is in preparation for the Ubuntu 18.04 discussion; it only clarifies the status quo. I'd still appreciate if you can have a look and approve this PR.

File: doc/ug/install_instructions/index.md [hunk 1]
```diff
@@ -4,30 +4,83 @@ title: Install Build Requirements
...
-* We develop and test on the following Linux distributions:
-  * Ubuntu 16.04 LTS
-  * Debian testing
-  * openSUSE Tumbleweed
-  * TODO: Check RHEL/CentOS and SLES (used in many commercial environments)
-
-TODO: Be more specific about the system requirements, especially the Linux distribution.
+* Physical access to that machine, root permissions and a graphical environment are assumed in this guide.
+  However, with an appropriate setup, none of these things are strictly needed.
+* 60 GB or more of disk space, depending on the used EDA tools.
```
Comment (rswarbrick) [inline doc/ug/install_instructions/index.md:14]: Grammar nit: This should be "on the EDA tools used".

File: doc/ug/install_instructions/index.md [hunk 2]
```diff
@@ -4,30 +4,83 @@ title: Install Build Requirements
...
+  (EDA tools like Xilinx Vivado can easily take up 40 GB each.)
+* Linux
+  * Ubuntu 16.04 LTS is the recommended reference platform.
+    Our continuous integration setup runs on Ubuntu 16.04 LTS, which gives us the most confidence that this distribution works out of the box.
+  * We do our best to support other Linux distributions.
+    However, we cannot guarantee they can be used "out of the box" and might require updates of packages.
+    Please file a [GitHub issue](https://github.com/lowRISC/opentitan/issues) if you need help or would like to propose a change to increase compatibility with other distributions.
+    * RedHat Enterprise Linux (RHEL)/CentOS 7 and 8
+    * Many of our contributors use rolling releaese distributions, which generally include the latest stable releases of all software packages, such as Debian unstable, openSUSE Tumbleweed, or Arch Linux.
```
Comment (rswarbrick) [inline doc/ug/install_instructions/index.md:23]: Nit: Typo in "release"

File: doc/ug/install_instructions/index.md [hunk 3]
```diff
@@ -4,30 +4,83 @@ title: Install Build Requirements
...
+* lsb-release
+* srecord
+* GNU make
+* pkg-config or pkgconf
+* libelf with development headers
+* libftdi with development headers
+* OpenSSL 1.0 or 1.1 with development headers
+* libusb 1.0
+* OpenOCD
+* [Verible](https://github.com/google/verible) {{< tool_version "verible" >}}
```
Comment (rswarbrick) [inline doc/ug/install_instructions/index.md:50]: I like the `tool_version` shortcode! Would it make sense to use it for `ninja` too?

**Review by rswarbrick — APPROVED** [review]
This looks good. Would it make sense to leave a note in our apt-requirements file to update the install instructions if we add a new dependency (or changed version)?

**Comment by imphil** [comment]
Thanks for having a look @rswarbrick. I updated the documentation with your comments, and also improved the wording around RHEL/CentOS.

File: apt-requirements.txt [hunk 4]
```diff
@@ -3,6 +3,8 @@
 # SPDX-License-Identifier: Apache-2.0
 
 # Keep sorted.
+# When updating this file also check if doc/ug/install_instructions/index.md
+# needs to be updated as well.
```
Comment (lenary) [inline apt-requirements.txt:7]: Is it worth pointing out that this list does get included in the install instructions where apt is mentioned?

File: doc/ug/install_instructions/index.md [hunk 5]
```diff
@@ -4,30 +4,89 @@ title: Install Build Requirements
...
+
+Note: Version numbers given below indicate a minimum version, newer versions are generally expected to work, unless explicitly stated.
+For packages listed below without a version number we have not determined a minimum version.
+
+* git
+* Python 3.5.2 with pip.
+  Additional Python dependencies are installed through pip.
+* A C++14 capable compiler.
+  GCC 5 or Clang 3.5 should meet this requirement.
+* clang-format
```
Comment (lenary) [inline doc/ug/install_instructions/index.md:42]: Somewhere in the docs, we say exactly what version is required. Right now, I believe it's `3.8.x`. 

I think the "somewhere" is in the `utils/container/README.md`, which might need merging with this?

**Review by lenary — COMMENTED** [review]
Some nits. Let's not duplicate this info between the documentation for the utilities container and the install instructions, which have quite a bit of overlap.

File: doc/ug/install_instructions/index.md (hunk 5 above)
Comment (imphil) [reply doc/ug/install_instructions/index.md:42]: We don't require a specific version at the moment, whatever is available as part of a supported LLVM version is fine. So effectively clang-format 3.5+ at the moment.

**Comment by imphil** [comment]
> Let's not duplicate this info between the documentation for the utilities container and the install instructions, which have quite a bit of overlap.

I updated the container README and also fixed some inconsistencies along the way, PTAL.

File: doc/ug/install_instructions/index.md (hunk 5 above)
Comment (lenary) [reply doc/ug/install_instructions/index.md:42]: Not true. More recent clang-format versions have subtle differences to the results of running clang-format in CI. See for instance #2035 and #2082. This is an issue googlers have run into, which is why a specific version of clang-format was documented at all.

**Review by lenary — APPROVED** [review]
LGTM modulo documenting the clang-format version.

File: doc/ug/install_instructions/index.md (hunk 5 above)
Comment (imphil) [reply doc/ug/install_instructions/index.md:42]: Can you point me to this documentation?

Yes, we use a specific version in CI and using another version will be painful on occasion, but I don't think we have a policy requiring a specific version at this point.

Comment (imphil) [reply doc/ug/install_instructions/index.md]: I added a sentence on clang-format 3.8 being recommended, please let me know if that addresses your concerns @lenary.

**Review by rasmus-madsen — APPROVED** [review]
LGTM

**Review by lenary — APPROVED** [review]

**Pull request merged**